### PR TITLE
[FLINK-15288] Starting jobmanager pod should respect containerized.heap-cutoff

### PR DESCRIPTION
**Pull request by wangyang0918**
## What is the purpose of the change

Starting jobmanager pod should respect containerized.heap-cutoff. The cutoff will be used to leave some memory for jvm off-heap, for example meta space, thread native memory and etc.

This ticket will not introduce new config options for cutoff, we will reuse the `ResourceManagerOptions#CONTAINERIZED_HEAP_CUTOFF_RATIO`. And we will keep the same behavior with Yarn/Mesos.

## Brief change log

* Move yarn Utils#calculateHeapSize to BootstrapTools
* Starting jobmanager pod should respect containerized.heap-cutoff

## Verifying this change

* The changes is covered by unit test `KubernetesUtilsTest#testGetJobManagerStartCommand`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by wangyang0918** [comment]
@TisonKun Could you please take a look this PR? It is a very small changes and need to be fixed in 1.10.

File: flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java [hunk 1]
```diff
@@ -56,14 +56,14 @@
 			" Its not possible to use this configuration key to define port ranges.");
 
 	/**
-	 * Percentage of heap space to remove from containers (YARN / Mesos), to compensate
+	 * Percentage of heap space to remove from containers (YARN / Mesos/ Kubernetes), to compensate
```
Comment (zhijiangW) [inline flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java]: Also need to adjust the descriptions in `ConfigConstants#CONTAINERIZED_HEAP_CUTOFF_RATIO`?

File: flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java [hunk 2]
```diff
@@ -56,14 +56,14 @@
...
 	 * for other JVM memory usage.
 	 */
 	public static final ConfigOption<Float> CONTAINERIZED_HEAP_CUTOFF_RATIO = ConfigOptions
```
Comment (zhijiangW) [inline flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java]: My previous thought for this option was for all the containers including `TaskExecutor`. But after reviewing the related codes I find it is only for master container. The `TaskExecutor` container would never use it.

So maybe we can rename this option to limit the scope to make it more clearly, but it is out of scope of this PR.

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesClusterDescriptorTest.java [hunk 3]
```diff
@@ -102,8 +102,8 @@ public void testKillCluster() throws Exception {
 		final KubernetesClusterDescriptor descriptor = new KubernetesClusterDescriptor(FLINK_CONFIG, flinkKubeClient);
 
 		final ClusterSpecification clusterSpecification = new ClusterSpecification.ClusterSpecificationBuilder()
-			.setMasterMemoryMB(1)
-			.setTaskManagerMemoryMB(1)
+			.setMasterMemoryMB(1234)
```
Comment (zhijiangW) [inline flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesClusterDescriptorTest.java]: We do not need to set these options, and the default value 768 in `ClusterSpecificationBuilder ` is already suitable for the test.

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesUtilsTest.java [hunk 4]
```diff
@@ -87,41 +88,43 @@ public void testGetJobManagerStartCommand() {
 		final Configuration cfg = new Configuration();
 
 		final String jmJvmOpts = "-DjmJvm";
+		final int jmMemoryAfterCutoff = BootstrapTools.calculateHeapSize(jobManagerMem, cfg);
+		final String jmJvmMem = String.format("-Xms%dm -Xmx%dm", jmMemoryAfterCutoff, jmMemoryAfterCutoff);
```
Comment (zhijiangW) [inline flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesUtilsTest.java]: We can still make these two vars as class level, and then `jmJvmMem` is not needed to pass into below `getJobManagerExpectedCommand` every time, because it is actually never changed.

File: flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java (hunk 1 above)
Comment (wangyang0918) [reply flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java]: Nice catch. I will also update the description of `ConfigConstants#CONTAINERIZED_HEAP_CUTOFF_RATIO`.

File: flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java (hunk 2 above)
Comment (wangyang0918) [reply flink-core/src/main/java/org/apache/flink/configuration/ResourceManagerOptions.java]: Yes, you are right. Currently, the cutoff is only effective for jobmanager. And i think we do need to rename it, because it will be deprecated soon. You could the discussion under https://issues.apache.org/jira/browse/FLINK-15288.

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesClusterDescriptorTest.java (hunk 3 above)
Comment (wangyang0918) [reply flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesClusterDescriptorTest.java]: I will fix it.

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesUtilsTest.java [hunk 5]
```diff
@@ -144,6 +147,15 @@ public void testGetJobManagerStartCommand() {
 				" " + mainClass + " " + mainClassArgs + " " + jmLogRedirects,
 			getJobManagerStartCommand(cfg, true, true, mainClassArgs));
 
+		cfg.setInteger(ResourceManagerOptions.CONTAINERIZED_HEAP_CUTOFF_MIN, 300);
```
Comment (zhijiangW) [inline flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesUtilsTest.java]: I think this part seems a bit reductant, because it tries to verify the internal implementation of `BootstrapTools#calculateHeapSize` which is not brought by this PR.
IMO, this utility method should be verified completely when introduced first time. If not, we should separately verify it in general, not only in kubernetes case.

I am not strongly against if you prefer to adding it here, but another concern is that can we somehow reuse `getJobManagerExpectedCommand` in `assertEquals`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java [hunk 6]
```diff
@@ -640,4 +641,30 @@ public static String getDynamicProperties(Configuration baseConfig, Configuratio
 			.toArray(String[]::new);
 		return String.join(" ", newAddedConfigs);
 	}
+
+	/**
+	 * Calculate heap size after cut-off.
+	 */
+	public static int calculateHeapSize(int memory, org.apache.flink.configuration.Configuration conf) {
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java]: I am not quite sure whether this function is out of scope of the class. From the definition of this class, it should not cover this function. Can you find other proper places for this utility or we can define a new util class for it?

Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java]: nit: can remove the prefix `org.apache.flink.configuration`

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesUtilsTest.java (hunk 5 above)
Comment (wangyang0918) [reply flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesUtilsTest.java]: I will remove this part of checking cutoff. Just as you say, it has already be checked by its own tests.

We always try to reuse `getJobManagerExpectedCommand`. However, when we change the `CONTAINER_START_COMMAND_TEMPLATE` and test the behavior, we could not reuse it. They do not share lots of common codes.

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesUtilsTest.java (hunk 4 above)
Comment (wangyang0918) [reply flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesUtilsTest.java]: Make sense. I will try to simply the codes.

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java (hunk 6 above)
Comment (wangyang0918) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java]: Currently, the only usage of calculating cutoff heap size is to set `-Xmx -Xms` for jobmanager start command. So i intend to update the description of this method to reflect this. When we refactor the memory configuration for jobmanager based on FLIP-49, we could  reconsider whether we need a separate class for jobmanager resource utils.

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java [hunk 7]
```diff
@@ -423,4 +425,60 @@ public void testGetDynamicProperties() {
...
+
+		Assert.assertEquals(616, BootstrapTools.calculateHeapSize(1000, conf));
+		Assert.assertEquals(8500, BootstrapTools.calculateHeapSize(10000, conf));
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void illegalArgument() {
+		final Configuration conf = new Configuration();
+		conf.setString(ResourceManagerOptions.CONTAINERIZED_HEAP_CUTOFF_RATIO.key(), "1.1");
+		Assert.assertEquals(0, BootstrapTools.calculateHeapSize(4000, conf));
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java]: nit: it may be strange to assert. We can use `BootstrapTools.calculateHeapSize(4000, conf)` directly here.

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java [hunk 8]
```diff
@@ -423,4 +425,60 @@ public void testGetDynamicProperties() {
...
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void illegalArgumentNegative() {
+		final Configuration conf = new Configuration();
+		conf.setString(ResourceManagerOptions.CONTAINERIZED_HEAP_CUTOFF_RATIO.key(), "-0.01");
+		Assert.assertEquals(0, BootstrapTools.calculateHeapSize(4000, conf));
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java]: ditto: remove `Assert.assertEquals`

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java [hunk 9]
```diff
@@ -423,4 +425,60 @@ public void testGetDynamicProperties() {
...
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void tooMuchCutoff() {
+		final Configuration conf = new Configuration();
+		conf.setString(ResourceManagerOptions.CONTAINERIZED_HEAP_CUTOFF_RATIO.key(), "6000");
+		Assert.assertEquals(0, BootstrapTools.calculateHeapSize(4000, conf));
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java]: ditto

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java (hunk 7 above)
Comment (wangyang0918) [reply flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java]: I will fix it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java (hunk 8 above)
Comment (wangyang0918) [reply flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java]: I will fix it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java (hunk 9 above)
Comment (wangyang0918) [reply flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java]: I will fix it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java [hunk 10]
```diff
@@ -423,4 +425,60 @@ public void testGetDynamicProperties() {
 		String dynamicProperties = BootstrapTools.getDynamicProperties(baseConfig, targetConfig);
 		assertEquals("-Dkey.b=b2 -Dkey.c=c", dynamicProperties);
 	}
+
+	/**
+	 * Remove 15% of the heap, at least 384MB.
+	 *
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java]: nit: remove one line

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java [hunk 11]
```diff
@@ -423,4 +425,60 @@ public void testGetDynamicProperties() {
 		String dynamicProperties = BootstrapTools.getDynamicProperties(baseConfig, targetConfig);
 		assertEquals("-Dkey.b=b2 -Dkey.c=c", dynamicProperties);
 	}
+
+	/**
+	 * Remove 15% of the heap, at least 384MB.
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java]: This description can not cover all the cases of this test. You can remove the description completely or give a more precise one.

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesUtilsTest.java (hunk 5 above)
Comment (zhijiangW) [reply flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesUtilsTest.java]: After we removing this new added part, we never have the reusing issue. And the previous ones were not in the scope of this PR. :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java (hunk 6 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java]: Make sense for me.

**Comment by wangyang0918** [comment]
@zhijiangW Thanks a lot for your detailed review. I have addressed all your comments. Please take a look again.

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java (hunk 10 above)
Comment (wangyang0918) [reply flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java]: Have fixed it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java (hunk 11 above)
Comment (wangyang0918) [reply flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/BootstrapToolsTest.java]: I have removed it.

**Comment by wangyang0918** [comment]
@zhijiangW Thanks for your kindly reminding. I will dive into the mesos deployment and create a follow-up ticket if needed.
